File: ports/zig/src/testing.zig
//! Render a view to an in-memory screen. No TTY, no escape codes, no timers —
//! which is what makes TUI code written with this library actually testable.
//!
//! ```zig
//! var screen = try testing.renderToScreen(allocator, 80, 24, "dark", ui.Body.plain(view));
//! defer screen.deinit();
//! try std.testing.expect(screen.contains("72%"));
//! ```
//!
//! A `RenderedScreen` owns an arena holding the framebuffer and everything the
//! frame allocated, so a test frees one thing rather than tracking a tree.

const std = @import("std");

const buffer_mod = @import("buffer.zig");
const capabilities_mod = @import("capabilities.zig");
const color_mod = @import("color.zig");
const diff = @import("diff.zig");
const surface_mod = @import("surface.zig");
const theme_mod = @import("theme.zig");
const ui = @import("ui.zig");
const unicode = @import("unicode.zig");

const Attrs = buffer_mod.Attrs;
const Capabilities = capabilities_mod.Capabilities;
const Color = color_mod.Color;
const FrameBuffer = buffer_mod.FrameBuffer;
const Surface = surface_mod.Surface;
const Theme = theme_mod.Theme;

pub const CellSnapshot = struct {
    /// The cell's text. Empty for the right half of a wide glyph.
    text: []const u8,
    fg: Color,
    bg: Color,
    attrs: Attrs,
};

pub const RenderedScreen = struct {
    width: usize,
    height: usize,
    buffer: FrameBuffer,
    theme: Theme,
    capabilities: Capabilities,
    /// Mouse regions the view registered, in draw order. Lets a test assert
    /// that a widget is actually reachable by the wheel or a click, which is
    /// otherwise only observable by running a real terminal.
    regions: []const ui.HitRegion,
    /// Focusable control ids, in registration order.
    focus_ids: []const []const u8,

    arena: *std.heap.ArenaAllocator,

    pub fn deinit(self: *RenderedScreen) void {
        const parent = self.arena.child_allocator;
        self.arena.deinit();
        parent.destroy(self.arena);
        self.* = undefined;
    }

    /// The arena the frame was rendered in. Anything it returns lives until
    /// `deinit`, so a test never has to free a row of text.
    pub fn allocator(self: RenderedScreen) std.mem.Allocator {
        return self.arena.allocator();
    }

    /// Plain text, one line per row, trailing spaces trimmed.
    pub fn text(self: RenderedScreen) ![]const u8 {
        return self.buffer.toText(self.allocator());
    }

    /// One row of plain text, trailing spaces trimmed.
    pub fn line(self: RenderedScreen, y: usize) ![]const u8 {
        const row = try self.buffer.rowTextAlloc(self.allocator(), y);
        return std.mem.trimEnd(u8, row, " ");
    }

    /// Everything, with ANSI colors — paste into a terminal to see it.
    pub fn ansi(self: RenderedScreen) ![]const u8 {
        const alloc = self.allocator();
        var empty = try FrameBuffer.init(alloc, self.width, self.height);
        defer empty.deinit();
        var encoder = diff.Encoder.init(alloc, .{ .colors = .truecolor });
        defer encoder.deinit();
        const result = try encoder.encode(&empty, &self.buffer, true);
        return alloc.dupe(u8, result.output);
    }

    pub fn cell(self: RenderedScreen, x: usize, y: usize) CellSnapshot {
        const i = self.buffer.index(x, y);
        const value = self.buffer.chars[i];
        // `cellText` borrows from the cluster table for a cluster and writes
        // into `buf` for a bare codepoint, so the scratch buffer has to outlive
        // the call — it is copied into the arena rather than returned by
        // pointer into a dead stack frame.
        var scratch: [4]u8 = undefined;
        const shown: []const u8 = if (value == unicode.continuation)
            ""
        else if (value == 0)
            " "
        else
            self.allocator().dupe(u8, unicode.cellText(value, &scratch)) catch " ";
        return .{
            .text = shown,
            .fg = self.buffer.fg[i],
            .bg = self.buffer.bg[i],
            .attrs = self.buffer.attrs[i],
        };
    }

    /// Column and row of the first occurrence, or null.
    pub fn find(self: RenderedScreen, needle: []const u8) ?struct { x: usize, y: usize } {
        const alloc = self.allocator();
        for (0..self.height) |y| {
            const row = self.buffer.rowTextAlloc(alloc, y) catch return null;
            const byte = std.mem.indexOf(u8, row, needle) orelse continue;
            // Report the column, not the byte offset: a row with a wide glyph
            // in it has more bytes than cells.
            const x = std.unicode.utf8CountCodepoints(row[0..byte]) catch byte;
            return .{ .x = x, .y = y };
        }
        return null;
    }

    pub fn contains(self: RenderedScreen, needle: []const u8) bool {
        return self.find(needle) != null;
    }
};

/// Force a capability rather than detecting it, exactly as `capabilities` does.
pub const Overrides = capabilities_mod.Overrides;

/// Render a view to an in-memory screen.
pub fn renderToScreen(
    parent: std.mem.Allocator,
    width: usize,
    height: usize,
    theme_name: []const u8,
    view: ui.Body,
) !RenderedScreen {
    return renderWith(parent, width, height, theme_name, 0, .{}, view);
}

/// The full form: pick the frame number and override capabilities.
pub fn renderWith(
    parent: std.mem.Allocator,
    width: usize,
    height: usize,
    theme_name: []const u8,
    frame: u64,
    overrides: Overrides,
    view: ui.Body,
) !RenderedScreen {
    return renderWithCollapse(parent, width, height, theme_name, frame, overrides, false, view);
}

/// Render with adjacent panel borders merged, as `App.Options.collapse_borders`
/// does for a running app. The caller owns the result.
pub fn renderCollapsedToText(
    parent: std.mem.Allocator,
    width: usize,
    height: usize,
    theme_name: []const u8,
    view: ui.Body,
) ![]u8 {
    var screen = try renderWithCollapse(parent, width, height, theme_name, 0, .{}, true, view);
    defer screen.deinit();
    return parent.dupe(u8, try screen.text());
}

/// As `renderToScreen`, with adjacent panel borders merged. The parity
/// fixtures cover both modes because collapsing changes the layout, not only
/// the glyphs.
pub fn renderCollapsedToScreen(
    parent: std.mem.Allocator,
    width: usize,
    height: usize,
    theme_name: []const u8,
    view: ui.Body,
) !RenderedScreen {
    return renderWithCollapse(parent, width, height, theme_name, 0, .{}, true, view);
}

fn renderWithCollapse(
    parent: std.mem.Allocator,
    width: usize,
    height: usize,
    theme_name: []const u8,
    frame: u64,
    overrides: Overrides,
    collapse_borders: bool,
    view: ui.Body,
) !RenderedScreen {
    const arena = try parent.create(std.heap.ArenaAllocator);
    errdefer parent.destroy(arena);
    arena.* = .init(parent);
    errdefer arena.deinit();
    const alloc = arena.allocator();

    const resolved = try alloc.create(Theme);
    resolved.* = theme_mod.resolve(theme_name);
    const caps = try alloc.create(Capabilities);
    caps.* = headlessCapabilities(overrides);

    var buffer = try FrameBuffer.init(alloc, width, height);
    buffer.clear(resolved.background, resolved.foreground);

    var ctx = ui.Ctx.init(alloc, resolved, caps, width, height);
    ctx.collapse_borders = collapse_borders;
    ctx.frame = frame;

    const root = Surface.root(&buffer, resolved);
    var container = ui.Container.init(root, &ctx, .column, .{});
    if (view.call) |f| try f(view.ctx, &container);
    try container.flush();
    for (ctx.overlays.items) |overlay| try overlay.draw(alloc, root);

    return .{
        .width = width,
        .height = height,
        .buffer = buffer,
        .theme = resolved.*,
        .capabilities = caps.*,
        .regions = ctx.hits.items,
        .focus_ids = ctx.focus_ids.items,
        .arena = arena,
    };
}

/// A headless render should look like a capable terminal, not like whatever is
/// running the test suite — otherwise a plot silently degrades to ASCII in CI.
fn headlessCapabilities(overrides: Overrides) Capabilities {
    const merged: Overrides = .{
        .tty = overrides.tty orelse true,
        .colors = overrides.colors orelse .truecolor,
        .unicode = overrides.unicode orelse true,
        .braille = overrides.braille orelse true,
        .mouse = overrides.mouse,
        .synchronized_output = overrides.synchronized_output,
    };
    return capabilities_mod.detect(merged, capabilities_mod.Env.empty, true);
}

/// Shorthand: render and return plain text. Ideal for snapshot tests. The
/// caller owns the result.
pub fn renderToText(
    parent: std.mem.Allocator,
    width: usize,
    height: usize,
    theme_name: []const u8,
    view: ui.Body,
) ![]u8 {
    var screen = try renderToScreen(parent, width, height, theme_name, view);
    defer screen.deinit();
    return parent.dupe(u8, try screen.text());
}

/// Render with ANSI colors, e.g. to write a demo screenshot to a file. The
/// caller owns the result.
pub fn renderToAnsi(
    parent: std.mem.Allocator,
    width: usize,
    height: usize,
    theme_name: []const u8,
    view: ui.Body,
) ![]u8 {
    var screen = try renderToScreen(parent, width, height, theme_name, view);
    defer screen.deinit();
    return parent.dupe(u8, try screen.ansi());
}

// --------------------------------------------------------------------- html

pub const HtmlOptions = struct {
    font_size: f64 = 14,
    padding: f64 = 16,
    class_name: []const u8 = "hqtui-screen",
    /// The default stack is ordered by box-drawing and Braille coverage.
    font_family: []const u8 = "ui-monospace,SFMono-Regular,Menlo,DejaVu Sans Mono," ++
        "Liberation Mono,Consolas,Segoe UI Symbol,monospace",
};

fn escapeHtml(out: *std.ArrayList(u8), alloc: std.mem.Allocator, text: []const u8) !void {
    for (text) |c| switch (c) {
        '&' => try out.appendSlice(alloc, "&amp;"),
        '<' => try out.appendSlice(alloc, "&lt;"),
        '>' => try out.appendSlice(alloc, "&gt;"),
        else => try out.append(alloc, c),
    };
}

fn escapeAttr(out: *std.ArrayList(u8), alloc: std.mem.Allocator, text: []const u8) !void {
    for (text) |c| switch (c) {
        '&' => try out.appendSlice(alloc, "&amp;"),
        '<' => try out.appendSlice(alloc, "&lt;"),
        '>' => try out.appendSlice(alloc, "&gt;"),
        '"' => try out.appendSlice(alloc, "&quot;"),
        '\'' => try out.appendSlice(alloc, "&#39;"),
        else => try out.append(alloc, c),
    };
}

fn writeCssColor(
    out: *std.ArrayList(u8),
    alloc: std.mem.Allocator,
    value: Color,
    fallback: []const u8,
) !void {
    if (value.isDefault()) {
        try out.appendSlice(alloc, fallback);
        return;
    }
    try out.print(alloc, "#{x:0>6}", .{value.raw() & 0xff_ffff});
}

/// Render to standalone HTML — a real screenshot of the UI, no terminal needed.
/// The caller owns the result.
pub fn renderToHtml(
    alloc: std.mem.Allocator,
    screen: RenderedScreen,
    options: HtmlOptions,
) ![]u8 {
    var bg_fallback: std.ArrayList(u8) = .empty;
    defer bg_fallback.deinit(alloc);
    try writeCssColor(&bg_fallback, alloc, screen.theme.background, "#000");

    var fg_fallback: std.ArrayList(u8) = .empty;
    defer fg_fallback.deinit(alloc);
    try writeCssColor(&fg_fallback, alloc, screen.theme.foreground, "#fff");

    var body: std.ArrayList(u8) = .empty;
    defer body.deinit(alloc);

    var run: std.ArrayList(u8) = .empty;
    defer run.deinit(alloc);

    var open = false;
    var run_fg: Color = .default;
    var run_bg: Color = .default;
    var run_attrs: Attrs = Attrs.none;

    for (0..screen.height) |y| {
        if (y > 0) try body.append(alloc, '\n');
        for (0..screen.width) |x| {
            const i = screen.buffer.index(x, y);
            const value = screen.buffer.chars[i];
            if (value == unicode.continuation) continue;

            const fg = screen.buffer.fg[i];
            const bg = screen.buffer.bg[i];
            const attrs = screen.buffer.attrs[i];
            const same = open and fg.raw() == run_fg.raw() and
                bg.raw() == run_bg.raw() and attrs.eql(run_attrs);
            if (!same) {
                if (open) try flushRun(
                    &body,
                    &run,
                    alloc,
                    run_fg,
                    run_bg,
                    run_attrs,
                    fg_fallback.items,
                    bg_fallback.items,
                );
                run_fg = fg;
                run_bg = bg;
                run_attrs = attrs;
                open = true;
            }
            var scratch: [4]u8 = undefined;
            const shown = if (value == 0) " " else unicode.cellText(value, &scratch);
            try run.appendSlice(alloc, shown);
        }
        if (open) {
            try flushRun(
                &body,
                &run,
                alloc,
                run_fg,
                run_bg,
                run_attrs,
                fg_fallback.items,
                bg_fallback.items,
            );
            open = false;
        }
    }

    // Every one of these is spliced into an attribute, so none may be trusted.
    // Numbers are bounded, and a font stack is reduced to the characters a font
    // stack can legitimately contain — escaping alone still lets `;` open a new
    // CSS property.
    const font_size = boundedNumber(options.font_size, 14);
    const padding = boundedNumber(options.padding, 16);

    var font: std.ArrayList(u8) = .empty;
    defer font.deinit(alloc);
    for (options.font_family) |c| {
        const keep = std.ascii.isAlphanumeric(c) or
            std.mem.indexOfScalar(u8, " ,._'-", c) != null;
        if (keep) try font.append(alloc, c);
    }

    var out: std.ArrayList(u8) = .empty;
    errdefer out.deinit(alloc);

    // One <pre> with newline-separated rows: wrapping each row in its own
    // element gives the browser licence to lay out lines independently, which
    // pulls box-drawing rules apart. A single text flow tiles the grid exactly.
    // line-height must be tight for the same reason.
    try out.appendSlice(alloc, "<pre class=\"");
    try escapeAttr(&out, alloc, options.class_name);
    try out.appendSlice(alloc, "\" style=\"background:");
    try out.appendSlice(alloc, bg_fallback.items);
    try out.appendSlice(alloc, ";color:");
    try out.appendSlice(alloc, fg_fallback.items);
    try out.print(alloc, ";padding:{d}px;font-size:{d}px;line-height:{d:.2}px;font-family:", .{
        padding,
        font_size,
        font_size * 1.18,
    });
    try escapeAttr(&out, alloc, font.items);
    try out.appendSlice(
        alloc,
        ";margin:0;overflow-x:auto;border-radius:8px;white-space:pre;" ++
            "font-variant-ligatures:none;-webkit-font-smoothing:antialiased\">",
    );
    try out.appendSlice(alloc, body.items);
    try out.appendSlice(alloc, "</pre>");
    return out.toOwnedSlice(alloc);
}

fn boundedNumber(value: f64, fallback: f64) f64 {
    if (std.math.isFinite(value) and value > 0 and value <= 1000) return value;
    return fallback;
}

fn flushRun(
    body: *std.ArrayList(u8),
    run: *std.ArrayList(u8),
    alloc: std.mem.Allocator,
    fg: Color,
    bg: Color,
    attrs: Attrs,
    fg_fallback: []const u8,
    bg_fallback: []const u8,
) !void {
    if (run.items.len == 0) return;
    try body.appendSlice(alloc, "<span style=\"color:");
    try writeCssColor(body, alloc, fg, fg_fallback);
    try body.appendSlice(alloc, ";background:");
    try writeCssColor(body, alloc, bg, bg_fallback);
    if (attrs.bold) try body.appendSlice(alloc, ";font-weight:700");
    if (attrs.dim) try body.appendSlice(alloc, ";opacity:.65");
    if (attrs.italic) try body.appendSlice(alloc, ";font-style:italic");
    if (attrs.underline) try body.appendSlice(alloc, ";text-decoration:underline");
    try body.appendSlice(alloc, "\">");
    try escapeHtml(body, alloc, run.items);
    try body.appendSlice(alloc, "</span>");
    run.clearRetainingCapacity();
}

File: ports/rust/src/testing.rs
//! Render a view to an in-memory screen. No TTY, no escape codes, no timers —
//! which is what makes TUI code written with this library actually testable.
//!
//! ```
//! use hqtui::testing::render_to_screen;
//! use hqtui::ui::Panel;
//!
//! let screen = render_to_screen(80, 24, "dark", |ui| {
//!     ui.panel(Panel::new().title("CPU"), |p| { p.text("72%"); });
//! });
//! assert!(screen.contains("72%"));
//! ```

use std::cell::RefCell;
use std::rc::Rc;

use crate::buffer::{Attrs, FrameBuffer};
use crate::capabilities::{detect_capabilities_in, Capabilities, CapabilityOverrides, ColorDepth};
use crate::color::Color;
use crate::diff::{Encoder, EncoderOptions};
use crate::layout::Direction;
use crate::surface::Surface;
use crate::theme::{resolve_theme, Theme};
use crate::ui::{Container, Ctx, HitRegion, Layout};
use crate::unicode::{cell_text, CONTINUATION};

#[derive(Clone, Debug, PartialEq, Eq)]
pub struct CellSnapshot {
    pub char: String,
    pub fg: Color,
    pub bg: Color,
    pub attrs: Attrs,
}

pub struct RenderedScreen {
    pub width: usize,
    pub height: usize,
    pub buffer: FrameBuffer,
    pub theme: Rc<Theme>,
    /// Mouse regions the view registered, in draw order. Lets a test assert
    /// that a widget is actually reachable by the wheel or a click, which is
    /// otherwise only observable by running a real terminal.
    pub regions: Vec<HitRegion>,
    /// Focusable control ids, in registration order.
    pub focus_ids: Vec<String>,
}

impl RenderedScreen {
    /// Plain text, one line per row, trailing spaces trimmed.
    pub fn text(&self) -> String {
        self.buffer.to_text()
    }

    /// One row of plain text.
    pub fn line(&self, y: usize) -> String {
        self.buffer.row_text(y).trim_end().to_string()
    }

    /// Everything, with ANSI colors — paste into a terminal to see it.
    pub fn ansi(&self) -> String {
        let empty = FrameBuffer::new(self.width, self.height);
        Encoder::new(EncoderOptions { colors: Some(ColorDepth::TrueColor), monochrome: false })
            .encode(&empty, &self.buffer, true)
            .output
    }

    pub fn cell(&self, x: usize, y: usize) -> CellSnapshot {
        let i = self.buffer.index(x, y);
        let value = self.buffer.chars[i];
        CellSnapshot {
            char: if value == CONTINUATION {
                String::new()
            } else if value == 0 {
                " ".to_string()
            } else {
                cell_text(value)
            },
            fg: self.buffer.fg[i],
            bg: self.buffer.bg[i],
            attrs: self.buffer.attrs[i],
        }
    }

    /// Row and column of the first occurrence, or `None`.
    pub fn find(&self, needle: &str) -> Option<(usize, usize)> {
        for y in 0..self.height {
            let row = self.buffer.row_text(y);
            if let Some(byte) = row.find(needle) {
                // Report the column, not the byte offset: a row with a wide
                // glyph in it has more bytes than cells.
                let x = row[..byte].chars().count();
                return Some((x, y));
            }
        }
        None
    }

    pub fn contains(&self, needle: &str) -> bool {
        (0..self.height).any(|y| self.buffer.row_text(y).contains(needle))
    }
}

/// Render a view to an in-memory screen.
pub fn render_to_screen<'v>(
    width: usize,
    height: usize,
    theme: &str,
    view: impl FnOnce(&mut Container<'v>),
) -> RenderedScreen {
    render_with(width, height, theme, 0, CapabilityOverrides::default(), view)
}

thread_local! {
    /// Set for the duration of `render_collapsed`. A thread local rather than
    /// another argument on five render functions, because collapsing is a
    /// property of a whole screen and tests are the only caller that needs to
    /// change it per render.
    static COLLAPSE: std::cell::Cell<bool> = const { std::cell::Cell::new(false) };
}

/// Render with adjacent panel borders merged, as `AppOptions::collapse_borders`
/// does for a running app.
pub fn render_collapsed_to_text<'v>(
    width: usize,
    height: usize,
    theme: &str,
    view: impl FnOnce(&mut Container<'v>),
) -> String {
    render_collapsed_to_screen(width, height, theme, view).text()
}

/// The same, returning the whole screen. Cell-level parity against the
/// TypeScript reference needs the buffer, not the text.
pub fn render_collapsed_to_screen<'v>(
    width: usize,
    height: usize,
    theme: &str,
    view: impl FnOnce(&mut Container<'v>),
) -> RenderedScreen {
    COLLAPSE.with(|flag| flag.set(true));
    let out = render_to_screen(width, height, theme, view);
    COLLAPSE.with(|flag| flag.set(false));
    out
}

/// The full form: pick the frame number and override capabilities.
pub fn render_with<'v>(
    width: usize,
    height: usize,
    theme_name: &str,
    frame: u64,
    overrides: CapabilityOverrides,
    view: impl FnOnce(&mut Container<'v>),
) -> RenderedScreen {
    let theme = Rc::new(resolve_theme(theme_name));
    let capabilities = Rc::new(headless_capabilities(overrides));

    let buffer = Rc::new(RefCell::new(FrameBuffer::new(width, height)));
    buffer.borrow_mut().clear(theme.background, theme.foreground);

    let mut ctx = Ctx::new(theme.clone(), capabilities, width, height);
    ctx.frame = frame;
    ctx.collapse_borders = COLLAPSE.with(|flag| flag.get());
    let ctx = Rc::new(ctx);

    let root = Surface::root(buffer.clone(), theme.clone());
    let mut container =
        Container::new(root.clone(), ctx.clone(), Direction::Column, &Layout::default());
    view(&mut container);
    container.flush();
    for overlay in ctx.take_overlays() {
        overlay.draw(&root);
    }

    let regions = ctx.hits();
    let focus_ids = ctx.focus_ids();
    let buffer = Rc::try_unwrap(buffer)
        .map(RefCell::into_inner)
        .unwrap_or_else(|rc| rc.borrow().clone());

    RenderedScreen { width, height, buffer, theme, regions, focus_ids }
}

/// A headless render should look like a capable terminal, not like whatever is
/// running the test suite — otherwise a plot silently degrades to ASCII in CI.
fn headless_capabilities(overrides: CapabilityOverrides) -> Capabilities {
    let merged = CapabilityOverrides {
        tty: Some(overrides.tty.unwrap_or(true)),
        colors: Some(overrides.colors.unwrap_or(ColorDepth::TrueColor)),
        unicode: Some(overrides.unicode.unwrap_or(true)),
        braille: Some(overrides.braille.unwrap_or(true)),
        mouse: overrides.mouse,
        synchronized_output: overrides.synchronized_output,
    };
    detect_capabilities_in(&merged, &Default::default(), true)
}

/// Shorthand: render and return plain text. Ideal for snapshot tests.
pub fn render_to_text<'v>(
    width: usize,
    height: usize,
    theme: &str,
    view: impl FnOnce(&mut Container<'v>),
) -> String {
    render_to_screen(width, height, theme, view).text()
}

/// Render with ANSI colors, e.g. to write a demo screenshot to a file.
pub fn render_to_ansi<'v>(
    width: usize,
    height: usize,
    theme: &str,
    view: impl FnOnce(&mut Container<'v>),
) -> String {
    render_to_screen(width, height, theme, view).ansi()
}

fn escape_html(text: &str) -> String {
    text.chars()
        .map(|c| match c {
            '&' => "&amp;".to_string(),
            '<' => "&lt;".to_string(),
            '>' => "&gt;".to_string(),
            other => other.to_string(),
        })
        .collect()
}

fn escape_attr(text: &str) -> String {
    text.chars()
        .map(|c| match c {
            '&' => "&amp;".to_string(),
            '<' => "&lt;".to_string(),
            '>' => "&gt;".to_string(),
            '"' => "&quot;".to_string(),
            '\'' => "&#39;".to_string(),
            other => other.to_string(),
        })
        .collect()
}

fn css_color(color: Color, fallback: &str) -> String {
    if color.is_default() {
        return fallback.to_string();
    }
    format!("#{:06x}", color.raw() & 0xff_ffff)
}

#[derive(Clone, Debug)]
pub struct HtmlOptions {
    pub font_size: f64,
    pub padding: f64,
    pub class_name: String,
    /// The default stack is ordered by box-drawing and Braille coverage.
    pub font_family: String,
}

impl Default for HtmlOptions {
    fn default() -> HtmlOptions {
        HtmlOptions {
            font_size: 14.0,
            padding: 16.0,
            class_name: "hqtui-screen".into(),
            font_family: "ui-monospace,SFMono-Regular,Menlo,'DejaVu Sans Mono','Liberation Mono',Consolas,'Segoe UI Symbol',monospace".into(),
        }
    }
}

/// Render to standalone HTML — a real screenshot of the UI, no terminal needed.
pub fn render_to_html(screen: &RenderedScreen, options: &HtmlOptions) -> String {
    let theme = &screen.theme;
    let bg_fallback = css_color(theme.background, "#000");
    let fg_fallback = css_color(theme.foreground, "#fff");
    let buffer = &screen.buffer;
    let mut rows: Vec<String> = Vec::new();

    for y in 0..screen.height {
        let mut row = String::new();
        let mut run = String::new();
        let mut run_fg: Option<Color> = None;
        let mut run_bg: Option<Color> = None;
        let mut run_attrs: Option<Attrs> = None;

        let flush = |row: &mut String,
                     run: &mut String,
                     fg: Option<Color>,
                     bg: Option<Color>,
                     attrs: Option<Attrs>| {
            if run.is_empty() {
                return;
            }
            let attrs = attrs.unwrap_or(Attrs::NONE);
            let mut style = format!(
                "color:{};background:{}",
                css_color(fg.unwrap_or(Color::DEFAULT), &fg_fallback),
                css_color(bg.unwrap_or(Color::DEFAULT), &bg_fallback)
            );
            if attrs.contains(Attrs::BOLD) {
                style.push_str(";font-weight:700");
            }
            if attrs.contains(Attrs::DIM) {
                style.push_str(";opacity:.65");
            }
            if attrs.contains(Attrs::ITALIC) {
                style.push_str(";font-style:italic");
            }
            if attrs.contains(Attrs::UNDERLINE) {
                style.push_str(";text-decoration:underline");
            }
            row.push_str(&format!("<span style=\"{style}\">{}</span>", escape_html(run)));
            run.clear();
        };

        for x in 0..screen.width {
            let i = buffer.index(x, y);
            if buffer.chars[i] == CONTINUATION {
                continue;
            }
            let fg = buffer.fg[i];
            let bg = buffer.bg[i];
            let attrs = buffer.attrs[i];
            if Some(fg) != run_fg || Some(bg) != run_bg || Some(attrs) != run_attrs {
                flush(&mut row, &mut run, run_fg, run_bg, run_attrs);
                run_fg = Some(fg);
                run_bg = Some(bg);
                run_attrs = Some(attrs);
            }
            let value = buffer.chars[i];
            run.push_str(&if value == 0 { " ".to_string() } else { cell_text(value) });
        }
        flush(&mut row, &mut run, run_fg, run_bg, run_attrs);
        rows.push(row);
    }

    // Every one of these is spliced into an attribute, so none may be trusted.
    // Numbers are bounded, and a font stack is reduced to the characters a font
    // stack can legitimately contain — escaping alone still lets `;` open a new
    // CSS property.
    let number = |value: f64, fallback: f64| -> f64 {
        if value.is_finite() && value > 0.0 && value <= 1000.0 {
            value
        } else {
            fallback
        }
    };
    let font_size = number(options.font_size, 14.0);
    let padding = number(options.padding, 16.0);
    let font: String = options
        .font_family
        .chars()
        .filter(|c| c.is_ascii_alphanumeric() || " ,._'-".contains(*c))
        .collect();

    // One <pre> with newline-separated rows: wrapping each row in its own
    // element gives the browser licence to lay out lines independently, which
    // pulls box-drawing rules apart. A single text flow tiles the grid exactly.
    // line-height must be tight for the same reason.
    format!(
        "<pre class=\"{}\" style=\"background:{bg_fallback};color:{fg_fallback};\
padding:{padding}px;font-size:{font_size}px;line-height:{:.2}px;font-family:{};\
margin:0;overflow-x:auto;border-radius:8px;white-space:pre;font-variant-ligatures:none;\
-webkit-font-smoothing:antialiased\">{}</pre>",
        escape_attr(&options.class_name),
        font_size * 1.18,
        escape_attr(&font),
        rows.join("\n")
    )
}
